TimeZone Mis-Match Between Start and End Times

<!--- This template provides sections for bugs and features. Please delete any irrelevant sections before submitting -->

#### Expected Behavior
<!--- Required. Describe the behavior you expect to see for your case. -->

This algorithm runs without error
```
class MyAlgorithm(QCAlgorithm):

    def initialize(self):
        self.set_start_date(2010, 1, 1)
        self.schedule.on(
            FuncDateRule(name="name", get_dates_function=self._get_dates),
            self.time_rules.at(0, 1),
            lambda: None
        )

    def _get_dates(self, start_date, end_date):
        start_date <= end_date
        return []
```

#### Actual Behavior
<!--- Required. Describe the actual behavior for your case. -->
We get
> [ERROR] FATAL UNHANDLED EXCEPTION:Engine.Run(): Error running algorithm Python.Runtime.PythonException: can't compare offset-naive and offset-aware datetimes, File "/QuantConnect/backtesting/./cache/algorithm/project/main.py", line 16, in _get_dates, start_date <= end_date, at Python.Runtime.PythonException.ThrowLastAsClrException() in src
untimePythonException.cs:line 52, at Python.Runtime.Dispatcher.TrueDispatch(Object[] args) in src
untimeDelegateManager.cs:line 346, at Python.Runtime.Dispatcher.Dispatch(Object[] args) in src
untimeDelegateManager.cs:line 210, at __System_Func`3[[System_DateTime, System_Private_CoreLib, Version=6_0_0_0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System_DateTime, System_Private_CoreLib, Version=6_0_0_0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System_Collections_Generic_IEnumerable`1[[System_DateTime, System_Private_CoreLib, Version=6_0_0_0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]

To avoid this error, I need to use 
```
start_date.replace(tzinfo=None) <= end_date
```
#### Potential Solution
<!--- Optional. Describe any potential solutions and/or thoughts as to what may be causing the difference between expected and actual behavior. -->
https://github.com/QuantConnect/Lean/blob/master/Common/Scheduling/ScheduleManager.cs#L284
Should be `Time.EndOfTime.ToUniversalTime()`

#### Reproducing the Problem
<!--- Required for Bugs. Describe how to reproduce the problem. This can be via a failing unit test or a simplified algorithm that reliably demonstrates this issue.  -->
Run algorithm above

#### System Information
<!--- Required for Bugs. Include any system specific information, such as OS. -->
QC Cloud

#### Checklist
<!--- Confirm that you've provided all the required information. -->
<!--- Required fields --->
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
<!--- Required for Bugs, feature request can delete the line below. -->
- [x] I have provided detailed steps to reproduce the issue

<!--- Template inspired by https://github.com/stevemao/github-issue-templates -->

--- a/Algorithm.Python/BaseFrameworkRegressionAlgorithm.py
+++ b/Algorithm.Python/BaseFrameworkRegressionAlgorithm.py
@@ -35,7 +35,7 @@ class BaseFrameworkRegressionAlgorithm(QCAlgorithm):
         # With this procedure, the Alpha Model will experience multiple universe changes
         self.add_universe_selection(ScheduledUniverseSelectionModel(
             self.date_rules.every_day(), self.time_rules.midnight,
-            lambda dt: symbols if dt.replace(tzinfo=None) < self.end_date - timedelta(1) else []))
+            lambda dt: symbols if dt < self.end_date - timedelta(1) else []))
 
         self.set_alpha(ConstantAlphaModel(InsightType.PRICE, InsightDirection.UP, timedelta(31), 0.025, None))
         self.set_portfolio_construction(EqualWeightingPortfolioConstructionModel())

--- a/Common/Data/UniverseSelection/ScheduledUniverse.cs
+++ b/Common/Data/UniverseSelection/ScheduledUniverse.cs
@@ -100,7 +100,7 @@ namespace QuantConnect.Data.UniverseSelection
         /// <returns>The data that passes the filter</returns>
         public override IEnumerable<Symbol> SelectSymbols(DateTime utcTime, BaseDataCollection data)
         {
-            return _selector(utcTime);
+            return _selector(DateTime.SpecifyKind(utcTime, DateTimeKind.Unspecified));
         }
 
         /// <summary>

--- a/Common/Scheduling/ScheduleManager.cs
+++ b/Common/Scheduling/ScheduleManager.cs
@@ -191,7 +191,7 @@ namespace QuantConnect.Scheduling
         {
             // back the date up to ensure we get all events, the event scheduler will skip past events that whose time has passed
             var dates = GetDatesDeferred(dateRule, _securities);
-            var eventTimes = timeRule.CreateUtcEventTimes(dates.Select(x => DateTime.SpecifyKind(x, DateTimeKind.Unspecified)));
+            var eventTimes = timeRule.CreateUtcEventTimes(dates);
             var scheduledEvent = new ScheduledEvent(name, eventTimes, callback);
             Add(scheduledEvent);
 

--- a/Common/Scheduling/ScheduleManager.cs
+++ b/Common/Scheduling/ScheduleManager.cs
@@ -20,6 +20,7 @@ using NodaTime;
 using QuantConnect.Securities;
 using QuantConnect.Logging;
 using Python.Runtime;
+using System.Linq;
 
 namespace QuantConnect.Scheduling
 {
@@ -190,7 +191,7 @@ namespace QuantConnect.Scheduling
         {
             // back the date up to ensure we get all events, the event scheduler will skip past events that whose time has passed
             var dates = GetDatesDeferred(dateRule, _securities);
-            var eventTimes = timeRule.CreateUtcEventTimes(dates);
+            var eventTimes = timeRule.CreateUtcEventTimes(dates.Select(x => DateTime.SpecifyKind(x, DateTimeKind.Unspecified)));
             var scheduledEvent = new ScheduledEvent(name, eventTimes, callback);
             Add(scheduledEvent);
 
@@ -281,7 +282,7 @@ namespace QuantConnect.Scheduling
         /// </summary>
         internal static IEnumerable<DateTime> GetDatesDeferred(IDateRule dateRule, SecurityManager securities)
         {
-            foreach (var item in dateRule.GetDates(securities.UtcTime.Date.AddDays(-1), Time.EndOfTime))
+            foreach (var item in dateRule.GetDates(DateTime.SpecifyKind(securities.UtcTime.Date.AddDays(-1), DateTimeKind.Unspecified), Time.EndOfTime))
             {
                 yield return item;
             }

--- a/Common/Scheduling/ScheduleManager.cs
+++ b/Common/Scheduling/ScheduleManager.cs
@@ -20,7 +20,6 @@ using NodaTime;
 using QuantConnect.Securities;
 using QuantConnect.Logging;
 using Python.Runtime;
-using System.Linq;
 
 namespace QuantConnect.Scheduling
 {

--- a/Tests/Common/Scheduling/ScheduleManagerTests.cs
+++ b/Tests/Common/Scheduling/ScheduleManagerTests.cs
@@ -23,6 +23,7 @@ using QuantConnect.Lean.Engine;
 using QuantConnect.Lean.Engine.DataFeeds;
 using QuantConnect.Lean.Engine.RealTime;
 using QuantConnect.Packets;
+using QuantConnect.Scheduling;
 using QuantConnect.Tests.Engine.DataFeeds;
 using QuantConnect.Util.RateLimit;
 
@@ -113,20 +114,8 @@ namespace QuantConnect.Tests.Common.Scheduling
         [Test]
         public void TriggersWeeklyScheduledEventsEachWeekLive()
         {
-            var algorithm = new AlgorithmStub();
-
-            var handler = new  TestableLiveTradingRealTimeHandler();
             var time = new DateTime(2024, 02, 10);
-            handler.ManualTimeProvider.SetCurrentTime(time);
-            var timeLimitManager = new AlgorithmTimeLimitManager(TokenBucket.Null, TimeSpan.FromMinutes(20));
-            handler.Setup(algorithm, new LiveNodePacket(), null, null, timeLimitManager);
-
-            algorithm.Schedule.SetEventSchedule(handler);
-
-            algorithm.SetDateTime(time);
-
-            var spy = algorithm.AddEquity("SPY").Symbol;
-
+            SetUp(time, out var algorithm, out var handler, out var spy);
             var eventTriggerTimes = new List<DateTime>();
             var scheduledEvent = algorithm.Schedule.On(algorithm.Schedule.DateRules.WeekStart(spy),
                 algorithm.Schedule.TimeRules.BeforeMarketClose(spy, 60),
@@ -172,6 +161,60 @@ namespace QuantConnect.Tests.Common.Scheduling
                 new DateTime(2024, 03, 25, 19, 0, 0),
             };
             CollectionAssert.AreEqual(expectedEventTriggerTimes, eventTriggerTimes);
+        }
+
+        [Test]
+        public void DatesReturnedAreNormalized()
+        {
+            var time = new DateTime(2024, 02, 10);
+            SetUp(time, out var algorithm, out var handler, out var spy);
+            var eventTriggerTimes = new List<DateTime>();
+            using var finished = new ManualResetEventSlim(false);
+
+            // Schedule a task to advance time
+            var timeStep = TimeSpan.FromMinutes(1);
+            var wasCalled = false;
+            Func<DateTime, DateTime, IEnumerable<DateTime>> func = (date1, date2) =>
+            {
+                Assert.AreEqual(DateTimeKind.Unspecified, date1.Kind);
+                Assert.AreEqual(DateTimeKind.Unspecified, date2.Kind);
+                wasCalled = true;
+                return new List<DateTime> { date1, date2 };
+            };
+
+            algorithm.Schedule.On(new FuncDateRule("Test", func),
+                algorithm.Schedule.TimeRules.Every(timeStep),
+                () =>
+                {
+                    handler.ManualTimeProvider.Advance(timeStep);
+                    var now = handler.ManualTimeProvider.GetUtcNow();
+                    finished.Set();
+                });
+
+            // Start
+            handler.SetTime(time);
+
+            finished.Wait(TimeSpan.FromSeconds(15));
+
+            handler.Exit();
+            Assert.IsTrue(wasCalled);
+        }
+
+        private void SetUp(DateTime time, out QCAlgorithm algorithm, out TestableLiveTradingRealTimeHandler handler, out Symbol spy)
+        {
+            algorithm = new AlgorithmStub();
+
+            handler = new TestableLiveTradingRealTimeHandler();
+
+            handler.ManualTimeProvider.SetCurrentTime(time);
+            var timeLimitManager = new AlgorithmTimeLimitManager(TokenBucket.Null, TimeSpan.FromMinutes(20));
+            handler.Setup(algorithm, new LiveNodePacket(), null, null, timeLimitManager);
+
+            algorithm.Schedule.SetEventSchedule(handler);
+
+            algorithm.SetDateTime(time);
+
+            spy = algorithm.AddEquity("SPY").Symbol;
         }
 
         private class TestableLiveTradingRealTimeHandler : LiveTradingRealTimeHandler

